This fixes an issue were a database session got stuck

stijndehaes opened the pull request:
The xcom_pull functionality got changed to be loaded lazy. This introduced a bug where a session is not properly closed. This only happens when inlets are defined on a task. It pops up on the airflow workers launched by the KubernetesExecutor. There is a database session stuck in idle in transaction for as long as the job is running. This uses precious resources.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

stijndehaes committed:
This fixes an issue were a database session got stuck

The xcom_pull functionality got changed to be loaded lazy.
This introduced a bug where a session is not properly closed.
This only happens when inlets are defined on a task.
It pops up on the airflow workers launched by the KubernetesExecutor.
There is a database session stuck in idle in transaction for as long
as the job is running. This uses precious resources.

uranusjr commented on airflow/lineage/__init__.py:
```diff
@@ -133,13 +134,16 @@ def wrapper(self, context, *args, **kwargs):
             # Remove auto and task_ids
             self.inlets = [i for i in self.inlets if not isinstance(i, str)]
-            _inlets = self.xcom_pull(context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS)
+            with create_session() as session:
+                _inlets = self.xcom_pull(
+                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS, session=session
+                )
```
Would be a good idea to add a comment explaining why this additional session creation is needed, and why it eliminates the issue (the dangling session in `LazyXComAccess`). It would also be a good idea to explain why the next two lines need to be indented into the context manager as well.

stijndehaes replied:
Done, I hope the explanation is clear?

uranusjr replied:
Maybe some more explaination on where the new session is created, and why it will not be properly closed? Otherwise the first intuition for anyone reading this would be to try to fix that unclosed session.

stijndehaes replied:
I update the explanation a bit. As to why the session is not properly closed, I actually do not know. Looking at the code again I can not figure out a reason why it would not close. In the code itself of `_get_bound_query` in the `LazyXComAccess` there is a try finally that closes the session.

uranusjr replied:
This is good enough I think. We don’t really know why the leak happens, but know this session does leak sometimes, for watever reason. Programming is hard.

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1456,12 +1456,14 @@ def xcom_push(
         dag_id: str | None = None,
         key: str = XCOM_RETURN_KEY,
         include_prior_dates: bool | None = None,
+        session: Session = NEW_SESSION,
```
```suggestion
        *,
        session: Session = NEW_SESSION,
```

stijndehaes replied:
Done

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1488,9 +1490,14 @@ def xcom_pull(
         :param include_prior_dates: If False, only XComs from the current
             execution_date are returned. If True, XComs from previous dates
             are returned as well.
+        :param session: the session used
```
Documenting the session argument is generally not needed.

stijndehaes replied:
Removed the session argument doc

ephraimbuddy commented on airflow/lineage/__init__.py:
```diff
@@ -133,13 +133,20 @@ def wrapper(self, context, *args, **kwargs):
+            # If we do not pass a session a new session will be created, however that session will not be
+            # properly closed and will remain open. After we are done iterating we can safely close this
+            # session.
+            with create_session() as session:
+                _inlets = self.xcom_pull(
+                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS, session=session
+                )
```
```suggestion
             _inlets = self.xcom_pull(
                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS
                )
```
The call to `create_session` is no longer needed since you decorated `xcom_pull` with `@provide_session`. The session will be automatically provided

ephraimbuddy replied:
Am i missing something @uranusjr ?

uranusjr replied:
I think both work (and using the default session from `@provide_session` is cleaner so you’re correct)

stijndehaes replied:
Removing the create_session will actually remove the bug fix. As explained in the comment above the create session: the issue is that xcom_pull returns a LazyXComAccess iterator. Since we have returned from the xcom_pull function, the session that would have been automatically created there will be closed and thus a new one that leaks will be started when we iterate. By wrapping this in a manually created session the leak is fixed

uranusjr replied:
Ah I see, the point is to make the session live slightly longer so line 147 below can still use it.

uranusjr replied:
Maybe a comment above this block would help future readers understand the context better. 

(Also I think the blank line + comments in lines 144-146 might actually hurt readability a bit since they make the two lines feel less associated. Maybe moving them above the entire block to make the `create_session`, `xcom_pull`, and`inlets.extend` lines stick close together would help as well.)

stijndehaes replied:
I remove the empty lines and 2 comments. I don't think the added much here.

There already is a comment above the create_session block to explain things, so I am not sure I correctly understand what you mean with `Maybe a comment above this block would help future readers understand the context better.`. Can you clarify it for me?

ephraimbuddy commented on airflow/models/baseoperator.py:
```diff
@@ -1456,12 +1456,15 @@ def xcom_push(
         dag_id: str | None = None,
         key: str = XCOM_RETURN_KEY,
         include_prior_dates: bool | None = None,
+        *,
```
```suggestion
```

uranusjr commented on airflow/lineage/__init__.py:
```diff
@@ -133,13 +133,20 @@ def wrapper(self, context, *args, **kwargs):
+                # xcom_pull returns a list of items for each given task_id
+                _inlets = [item for item in itertools.chain.from_iterable(_inlets)]
+
+                self.inlets.extend(_inlets)
```
```suggestion
                self.inlets.extend(i for i in itertools.chain.from_iterable(_inlets))
```

don’t need the intermediate list, I think

stijndehaes committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>

stijndehaes committed:
Make comment clearer

Remove the newlines and small comments. Should make it
clear that the comment above the block is really important

eladkal committed:
Merge branch 'main' into bugfix/session-not-properly-closed
